adagucserverEC/CCairoPlotter.cpp

```diff
@@ -1044,3 +1078,129 @@ void CCairoPlotter::writeToWebP32Stream(FILE *fp, unsigned char, int quality) {
 }
 
 #endif
+
+void CCairoPlotter::drawBarb(int x, int y, double direction, double strength, CColor barbColor, CColor outlineColor, bool drawOutline, float lineWidth, bool toKnots, bool flip) {
+
+  // Preserve path
+  cairo_save(cr);
+  cairo_path_t *cp = cairo_copy_path(cr);
+  cairo_new_path(cr);
+
+  // Draw
+  double wx1, wy1, wx2, wy2, dx1, dy1;
+  int strengthInKnots = round(strength);
+  if (toKnots) {
+    strengthInKnots = round(strength * 3600 / 1852.);
+  }
+
+  if (strengthInKnots <= 2) {
+    // Low wind, only draw a circle
+    cairo_arc(cr, x, y, 6, 0, 2 * M_PI);
+
+  } else {
+
+    int shaftLength = 30;
+
+    int nPennants = strengthInKnots / 50;
+    int nBarbs = (strengthInKnots % 50) / 10;
+    int rest = (strengthInKnots % 50) % 10;
+    int nhalfBarbs;
+    if (rest <= 2) {
+      nhalfBarbs = 0;
+    } else if (rest <= 7) {
+      nhalfBarbs = 1;
+    } else {
+      nhalfBarbs = 0;
+      nBarbs++;
+    }
```

Comment:
A bit scary to touch this code, but this seems to work for the cases around 50 as well, and is simpler:
```
        int strengthInFiveKnots = (strengthInKnots + 2.5) / 5;
        int nPennants2 = strengthInFiveKnots / 10;
        int nBarbs2 = (strengthInFiveKnots % 10) / 2;
        int nhalfBarbs2 = strengthInFiveKnots % 2;
```

This test code gives me some confidence:
```
#include <iostream>

using namespace std;

int main()
{
    for(int i=0; i <=200; i++) {
        int strengthInKnots = i;
        
        // New
        int strengthInFiveKnots = (strengthInKnots + 2.5) / 5;
        int nPennants2 = strengthInFiveKnots / 10;
        int nBarbs2 = (strengthInFiveKnots % 10) / 2;
        int nhalfBarbs2 = strengthInFiveKnots % 2;

//        printf("%d: %d %d %d\n", i, nPennants2, nBarbs2, nhalfBarbs2);

        // Old
        int nPennants = strengthInKnots / 50;
        int nBarbs = (strengthInKnots % 50) / 10;
        int rest = (strengthInKnots % 50) % 10;
        int nhalfBarbs;
        if (rest <= 2) {
          nhalfBarbs = 0;
        } else if (rest <= 7) {
          nhalfBarbs = 1;
        } else {
          nhalfBarbs = 0;
          nBarbs++;
        }
    
//        printf("%d: %d %d %d\n\n", i, nPennants, nBarbs, nhalfBarbs);
        if (nPennants != nPennants2 || nBarbs != nBarbs2 || nhalfBarbs != nhalfBarbs2) {
          printf("%d new : %d %d %d\n", i, nPennants2, nBarbs2, nhalfBarbs2);
          printf("%d old: %d %d %d\n\n", i, nPennants, nBarbs, nhalfBarbs);
            
        }
    }
    return 0;
}
```
        
Output:
```
48 new : 1 0 0
48 old: 0 5 0

49 new : 1 0 0
49 old: 0 5 0

98 new : 2 0 0
98 old: 1 5 0

99 new : 2 0 0
99 old: 1 5 0

148 new : 3 0 0
148 old: 2 5 0

149 new : 3 0 0
149 old: 2 5 0

198 new : 4 0 0
198 old: 3 5 0

199 new : 4 0 0
199 old: 3 5 0
```